libkbfs/bserver_remote.go

```diff
@@ -369,17 +369,33 @@ func (b *BlockServerRemote) batchDowngradeReferences(ctx context.Context,
 			if !tmpErr {
 				finalError = err
 				bid, err := BlockIDFromString(res.Failed.Bid.BlockHash)
+				found := false
 				if err == nil {
 					if refs, ok := contexts[bid]; ok {
 						for i := range refs {
 							if refs[i].GetRefNonce() == BlockRefNonce(res.Failed.Nonce) {
 								refs = append(refs[:i], refs[i+1:]...)
 								contexts[bid] = refs
+								// It's possible that the failure
+								// happened due to a complete network
+								// error.  If the zero block ref
+								// happens to be in the list, then we
+								// might end up making the call one
+								// more time since we can't tell if
+								// res.Failed.Nonce was initialized or
+								// not.
+								found = true
 								break
 							}
 						}
 					}
 				}
+				if !found {
+					// This wasn't a failure due to a particular
+					// reference, it was a complete failure (e.g., the
+					// network connection is down).
+					return nil, finalError
+				}
 			}
 		}
 
```

Comment:
If we haven't already made progress (i.e., we haven't removed a failed ref above), we may also want to make sure that the loop below makes some progress, i.e. we added at least one new nonce to the completed list, and return an error if not. That would be due to a buggy server, though, so it may not be that important to do it in this PR.